### ISSUE TITLE
[Question] Anchor optimization

### ISSUE THREAD

**Issue by vankhoa21991** [MIC-DKFZ/nnDetection#55]
### :question: Question

Hello, could you explain a bit further what happens inside the method find_anchor of the class BoxC001 in nndet/planning/architecture/boxes/base? 

As I understand you are optimizing the anchor size using the normalized bbox in the training set, but the method is not clear for me.

Thank you.

**Comment by mibaumgartner** [comment]
Dear @vankhoa21991 ,

first, all the training boxes are all centered around the origin (since the exact position of them during training will vary due to the patch extraction) and the method aims to optimize the anchors sizes. The anchors sizes are initialized with a fixed default value (this might be improvable) and then optimized by a black box optimizer to maximize the best anchor for each GT. To correctly reflect the Feature Pyramid Networks with its multiple levels, the current set of anchor boxes is extended by scaling them with the individual strides of the network (as is done by the anchor generator). The current default black box optimizer is based on Differential Evolution (I tried some others as well but found this one yields the best and most consistent results), more information can be collected from the `nevergrad` library. To allow anchors to specialize on individual "object types" only the best fitting anchor per object will be used to compute the final iou score.

So in Summary:
|---------------------------<<-------------------- optimize params ------------------------<<--------------------------|
Set of Anchors ---> Scale with Stride for Pyramid --->  IoU(Anchors, GT) ---> Select best fitting Anchor per GT as Score |


Best,
Michael

**Comment by vankhoa21991** [comment]
Thanks for your explanation, it's very helpful.

Van Khoa

**Comment by mibaumgartner** [comment]
I'll close the Issue for now, please feel free to reopen if any other questions come up.
